packages/core/contracts/oracle/implementation/VotingV2.sol

```diff
@@ -1071,6 +1110,122 @@ contract VotingV2 is
+     * @param identifier the price identifier to cancel the signal on.
+     * @param time the action time to cancel the signal on.
+     * @param ancillaryData the ancillary data to cancel the signal on.
+     */
+    function cancelSignalOnEmergencyAction(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public {
+        EmergencyAction storage action = emergencyActions[_encodePriceRequest(identifier, time, ancillaryData)];
+
+        require(action.accountSignaled[msg.sender] > 0);
+
+        // Decrement the cumulative signaled amount by the amount the user signed with in the beginning. We use this
+        // rather than using their activeStake as they may have changed this during the time they were signaled.
+        action.cumulativeSignaled -= action.accountSignaled[msg.sender];
+        action.accountSignaled[msg.sender] = 0; // Reset the signaled amount to 0.
+        stakerEmergencyActionSignalCount[msg.sender]--; // Decrement the num of actions this staker has signaled.
+
+        // Reset the blocking trackers used to prevent the staker from unstaking iff they have canceled all their
+        // signaled requests. Note that we do the equal and opposite operators of stakerEmergencyActionSignalCount
+        // to revert to their original values(if the staker had a pendingUnstake this should be recovered).
+        if (stakerEmergencyActionSignalCount[msg.sender] == 0) {
+            voterStakes[msg.sender].pendingUnstake -= 1;
+            voterStakes[msg.sender].unstakeRequestTime -= 10000000000;
+        }
+
+        emit SignaledOnEmergencyAction(
+            identifier,
+            time,
+            ancillaryData,
+            msg.sender,
+            stakerEmergencyActionSignalCount[msg.sender],
+            action.accountSignaled[msg.sender],
+            action.cumulativeSignaled
+        );
+    }
+
+    /**
+     * @notice Execute an emergency action by pushing the price 1e18 into the DVM for the given identifier, time and
+     * ancillary data. Note that emergencyActionThreshold needs to have singled on this action for it to be executed.
+     * @param identifier the price identifier to be executed and pushed into the DVMs resolved prices.
+     * @param time the action time to be executed and pushed into the DVMs resolved prices.
+     * @param ancillaryData the ancillary data to be executed and pushed into the DVMs resolved prices.
+     */
+    function executeEmergencyAction(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public {
+        bytes32 requestIdentifier = _encodePriceRequest(identifier, time, ancillaryData);
+        EmergencyAction storage action = emergencyActions[requestIdentifier];
+        require(action.executed == false);
+        require((action.cumulativeSignaled * 1e18) / getCumulativeStake() >= emergencyActionThreshold);
```

Comment:
This has some of the same issues as measuring the GAT against total staked. If there was a small amount staked, emergency actions could become very cheap. Thoughts on having a parameterizable _number_ of tokens rather than a percentage?

Reply:
Can staking ever get blocked?

Reply:
Yes. If update cumulative slashing trackers is reverting then you can get stuck.

Reply:
@md0x agreed. We face this risk on _every_ migration, whereby the first staker has full control of the protocol. I don't think this is an acceptable amount of risk.

We could use the GAT for this reasonableTokenAmount, so basically the minimum to push a vote through is the same as the min to push an emergency action through.

Thinking more about this, I think short term capital makes this really risky. I think maybe we'd want the following features to mitigate some of the risks:
1. The ability to vote _against_.
2. A long (1 week+) delay between reaching the threshold and execution whereby dissenters have the ability to vote against and push it back below the threshold.

Overall, though, I'm beginning to think the risks of a hostile attack/takeover may outweigh the rewards with a feature like this. I would really like to hear OZ's opinion on _this PR_ as a separate entity from the audit to see if they think this mitigates risk or makes things riskier.

Thoughts?

Reply:
Oh! You are right @mrice32, reentrancy guard doesn't help to prevent a flashloan attack in this case.

Reply:
Small late comment:To prevent flash loan attacks it might work to force signalOnEmergencyAction and executeEmergencyAction to be executed in separate blocks.